src/qcodes/instrument_drivers/CopperMountain/_M5xxx.py

```diff
@@ -0,0 +1,898 @@
+            s21 magnitude [dB], s21 phase [rad],
+            s22 magnitude [dB], s22 phase [rad]
+
+        """
+
+        with self.timeout.set_to(max(self.timeout(), expected_measurement_duration)):
+            self.write("CALC1:PAR:COUN 4")  # 4 trace
+            self.write("CALC1:PAR1:DEF S11")  # Choose S11 for trace 1
+            self.write("CALC1:PAR2:DEF S12")  # Choose S12 for trace 2
+            self.write("CALC1:PAR3:DEF S21")  # Choose S21 for trace 3
+            self.write("CALC1:PAR4:DEF S22")  # Choose S22 for trace 4
+            self._set_trace_formats_to_smith(traces=[1, 2, 3, 4])
+            self.write("TRIG:SEQ:SING")  # Trigger a single sweep
+            self.ask("*OPC?")  # Wait for measurement to complete
+
+            # Get data as string
+            freq_raw = self.ask("SENS1:FREQ:DATA?")
+            s11_raw = self.ask("CALC1:TRAC1:DATA:FDAT?")
+            s12_raw = self.ask("CALC1:TRAC2:DATA:FDAT?")
+            s21_raw = self.ask("CALC1:TRAC3:DATA:FDAT?")
+            s22_raw = self.ask("CALC1:TRAC4:DATA:FDAT?")
+
+        # Get data as numpy array
+        freq = np.fromstring(freq_raw, dtype=float, sep=",")
+        s11 = np.fromstring(s11_raw, dtype=float, sep=",")
+        s11 = s11[0::2] + 1j * s11[1::2]
+        s12 = np.fromstring(s12_raw, dtype=float, sep=",")
+        s12 = s12[0::2] + 1j * s12[1::2]
+        s21 = np.fromstring(s21_raw, dtype=float, sep=",")
+        s21 = s21[0::2] + 1j * s21[1::2]
+        s22 = np.fromstring(s22_raw, dtype=float, sep=",")
+        s22 = s22[0::2] + 1j * s22[1::2]
+
+        return (
+            np.array(freq),
+            self._db(s11),
+            np.array(np.angle(s11)),
+            self._db(s12),
+            np.array(np.angle(s12)),
+            self._db(s21),
+            np.array(np.angle(s21)),
+            self._db(s22),
+            np.array(np.angle(s22)),
+        )
+
+    def update_lin_traces(self) -> None:
+        """
+        Updates start, stop and number_of_points of all trace parameters so that the
+        setpoints and shape are updated for the sweep.
+        """
+        start = self.start()
+        stop = self.stop()
+        number_of_points = self.number_of_points()
+        for _, parameter in self.parameters.items():
+            if isinstance(parameter, (FrequencySweepMagPhase)):
+                try:
+                    parameter.set_sweep(start, stop, number_of_points)
+                except AttributeError:
+                    pass
+
+    def reset_averages(self) -> None:
+        """
+        Resets average count to 0
+        """
+        self.write("SENS1.AVER.CLE")
+
+    @staticmethod
+    def _db(data: np.ndarray) -> np.ndarray:
+        """
+        Return dB from magnitude
+
+        Args:
+            data: data to be transformed into dB.
+
+        Returns:
+            data: data transformed in dB.
+
+        """
+
+        return 20.0 * np.log10(np.abs(data))
+
+
+class FrequencySweepMagPhase(MultiParameter):
```

Comment:
For info only:
I consider the MultiParameter class to be Legacy, and many of the newer structures do not necessarily handle it in a new way. Instead, I would recommend returning a complex value or returning the Magnitude and Phase sweeps individually.

I see that this is how the Rohde-Schwarz ZNB-class VNA's implement their mag/phase sweep, so I'm not  asking for this to be changed.